### PR TITLE
Close race condition when allocating new block

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

Fixes https://github.com/projectcalico/libcalico-go/issues/1336

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Close race condition that could result in an extra IPAM block being allocated to a node.
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
I've had this fix running the flaky FV test from https://github.com/projectcalico/libcalico-go/issues/1336 for a while now with no failures. The code without this fix seemed to flake at least 1 in 15 times, sometimes quite a bit more.

File: lib/ipam/ipam.go [hunk 1]
```diff
@@ -486,9 +486,11 @@ func (s *blockAssignState) findOrClaimBlock(ctx context.Context, minFreeIps int)
 			// Claim a new block.
 			logCtx.Infof("No more affine blocks, but need to claim more blocks -- allocate another block")
 
-			// First, try to find an unclaimed block.
-			logCtx.Info("Looking for an unclaimed block")
-			subnet, err := s.client.blockReaderWriter.findUnclaimedBlock(ctx, s.host, s.version, s.pools, *config)
+			// First, try to find a usable block. findUsableBlock will usually return a new block, or in rare scenarios an already
+			// allocated affine block. This may happen due to a race condition where another process on the host allocates a new block
+			// after we decide that a new block is required to satisfy this request, but before we actually allocate a new block.
+			logCtx.Info("Looking for an affine block with space, or a new unclaimed block")
```
Comment (robbrockbank) [inline lib/ipam/ipam.go:492]: not: we can probably remove the previous log at line 487.

File: lib/ipam/ipam_block_reader_writer.go [hunk 2]
```diff
@@ -120,11 +128,15 @@ func (rw blockReaderWriter) findUnclaimedBlock(ctx context.Context, host string,
...
+			if info, ok := exists[subnet.String()]; !ok {
 				log.Infof("Found free block: %+v", *subnet)
 				return subnet, nil
+			} else if info.affinity == host && info.numFree != 0 {
+				// Belongs to this host and has free allocations.
+				log.Debugf("Block %s already assigned to host, has free space", subnet.String())
+				return subnet, nil
 			}
-			log.Debugf("Block %s already exists", subnet.String())
+			log.Debugf("Block %s already exists, try another", subnet.String())
```
Comment (robbrockbank) [inline lib/ipam/ipam_block_reader_writer.go:139]: really minor nit:  Might be nice to say in the log that assigned to different host or no free space.

**Review by robbrockbank — APPROVED** [review]

**Review by mgleung — APPROVED** [review]
LGTM

**Pull request merged**